In [1]:
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import MinMaxScaler
%matplotlib inline
import matplotlib.pyplot as plt
import random

In [2]:
#define paths

data_dir = '/analysis/shikhar/abcd_npc/img_dt/' #'/analysis/shikhar/abcd_npc/img_dt/' #or /data/Ritter/abcd_np/
trainy_filename = 'train/train_y.csv'
valy_filename = 'val/val_y.csv'
testy_filename ='test/test_y.csv'
trainx_filename = 'train/volumetrix_data_train.csv'
valx_filename = 'val/volumetrix_data_val.csv'
testx_filename = 'test/volumetrix_data_test.csv'

In [3]:
#loading datasets

trainy= pd.read_csv(data_dir+trainy_filename, index_col=False)
valy= pd.read_csv(data_dir+valy_filename, index_col=False)
testy= pd.read_csv(data_dir+testy_filename, index_col=False)
trainx =  pd.read_csv(data_dir+trainx_filename, index_col=False)
valx =  pd.read_csv(data_dir+valx_filename, index_col=False)
testx =  pd.read_csv(data_dir+testx_filename, index_col=False)

In [4]:
#append all

all_data = trainx.append(valx).append(testx).copy()
all_data['sex_bin'] = np.where(all_data['GENDER']=='F', 1, 0)
all_data = all_data.drop(['BTSV01_ID','DATASET_ID','GENDER','SRC_SUBJECT_ID','INTERVIEW_DATE'],\
                         axis=1).copy() ## removing irrelevant columns
trainyValy_moredt = pd.merge(all_data[['SUBJECTKEY','INTERVIEW_AGE','sex_bin','STUDY_COHORT_NAME']], \
                             trainy.append(valy), left_on='SUBJECTKEY', right_on='subject')
print(all_data.STUDY_COHORT_NAME.value_counts())
print(trainy.shape)
print(valy.shape)

ABCD NP Challenge 2019 Training Set      3736
ABCD NP Challenge 2019 Test Set          3640
ABCD NP Challenge 2019 Validation Set     415
Name: STUDY_COHORT_NAME, dtype: int64
(3739, 2)
(415, 2)


#### ^^^ Training set should have been 3739, test should have been 3648 <<< In order to use above we would need imputation
##### checked in R> all IDs in volumetrix data have scan data (i.e no new random ID)


----
### Adding missing subjects (no volumetrix data) and imputing their volumetrix values

In [5]:
# appending missing Subjects

missing_dict={'test':list(set(testy.subject).difference(set(testx.SUBJECTKEY))),\
         'train':list(set(trainy.subject).difference(set(trainx.SUBJECTKEY)))}

for key, values in missing_dict.items():
    if key=='test':
        studyname = 'ABCD NP Challenge 2019 Test Set'
        for key in values:
            all_data = all_data.append([{'SUBJECTKEY':key, 'STUDY_COHORT_NAME':studyname}], ignore_index=True)
    else:
        studyname = 'ABCD NP Challenge 2019 Training Set'
        for key in values:
            all_data = all_data.append([{'SUBJECTKEY':key, 'STUDY_COHORT_NAME':studyname}], ignore_index=True)

/home/shikhar/.local/lib/python3.6/site-packages/pandas/core/indexing.py:1472: FutureWarning: 
Passing list-likes to .loc or [] with any missing label will raise
KeyError in the future, you can use .reindex() as an alternative.

See the documentation here:
https://pandas.pydata.org/pandas-docs/stable/indexing.html#deprecate-loc-reindex-listlike
  return self._getitem_tuple(key)


In [6]:
print(all_data['sex_bin'].value_counts())
print(all_data['sex_bin'].isna().value_counts())

0.0    4085
1.0    3706
Name: sex_bin, dtype: int64
False    7791
True       11
Name: sex_bin, dtype: int64


In [7]:
#randomly assigning sex to appended subjects (only 11 new subjects)
np.random.seed(seed=234)
all_data['sex_bin'] = all_data['sex_bin'].fillna( \
                pd.Series(np.random.choice(all_data[all_data['sex_bin'].notna()]['sex_bin'], size=len(all_data.index)))\
                                                ) 
print(all_data['sex_bin'].value_counts())
print(all_data['sex_bin'].isna().value_counts())


0.0    4092
1.0    3710
Name: sex_bin, dtype: int64
False    7802
Name: sex_bin, dtype: int64


In [8]:
#imputing mean for rest of the columns (only 11 new subjects)
all_data = all_data.fillna(all_data.mean()).copy()
print(all_data.STUDY_COHORT_NAME.value_counts())

ABCD NP Challenge 2019 Training Set      3739
ABCD NP Challenge 2019 Test Set          3648
ABCD NP Challenge 2019 Validation Set     415
Name: STUDY_COHORT_NAME, dtype: int64


---
### Normalizing and checking for correlation

In [9]:
idv_scaler = MinMaxScaler(feature_range=(0, 1))
idv_scaler.fit(all_data[all_data.columns.difference(['SUBJECTKEY','STUDY_COHORT_NAME','sex_bin'])])
all_data[all_data.columns.difference(['SUBJECTKEY','STUDY_COHORT_NAME','sex_bin'])] = \
    idv_scaler.transform(all_data[all_data.columns.difference(['SUBJECTKEY','STUDY_COHORT_NAME','sex_bin'])])
all_data

SUBJECTKEY  INTERVIEW_AGE  SRI24PRECENTRALLGM  SRI24PRECENTRALRGM  \
0     NDAR_INV47FFMT85       0.730769            0.427935            0.312790   
1     NDAR_INVGP0KJLU3       0.653846            0.375488            0.484359   
2     NDAR_INVPMHGZU3M       0.500000            0.448786            0.504869   
3     NDAR_INVRPYKZRFL       0.384615            0.330454            0.468127   
4     NDAR_INVB8UX2D6Y       0.500000            0.318695            0.586295   
5     NDAR_INVXHD5E4MN       0.076923            0.238062            0.302160   
6     NDAR_INVLGBWLRM4       0.307692            0.152099            0.259396   
7     NDAR_INVUCGJ1U8F       0.500000            0.295462            0.496603   
8     NDAR_INVAVR6TURT       0.269231            0.296849            0.446853   
9     NDAR_INV930295U4       0.384615            0.484346            0.455493   
10    NDAR_INVH9G32LJW       0.807692            0.287294            0.318615   
11    NDAR_INV6ZU9NKBV       0.884615            0.407458            0.458383   
12    NDAR_INV96WYY4C5       0.538462            0.303177            0.505527   
13    NDAR_INVKLLP08KA       0.346154            0.371053            0.447592   
14    NDAR_INVFV0JWJ60       0.653846            0.422754            0.447958   
15    NDAR_INVXHKRJC1C       0.730769            0.417545            0.445879   
16    NDAR_INVHVXDEAGU       0.807692            0.486391            0.429055   
17    NDAR_INV2NZV8ELL       0.884615            0.438085            0.427309   
18    NDAR_INV36DKMERA       0.076923            0.458305            0.403826   
19    NDAR_INV1ZY3ZR47       0.923077            0.421483            0.519997   
20    NDAR_INV2V8997BF       0.346154            0.329778            0.413099   
21    NDAR_INVGALZWY31       0.692308            0.380261            0.314055   
22    NDAR_INVWZRB4WLL       0.038462            0.304608            0.350116   
23    NDAR_INV4T2VAM3Y       0.230769            0.300653            0.443500   
24    NDAR_INVLEXJMGJ6       0.307692            0.285837            0.490100   
25    NDAR_INVN1VPLV83       0.230769            0.550152            0.511187   
26    NDAR_INVWE1Z3L3H       0.346154            0.271181            0.766344   
27    NDAR_INVPFC26GTC       0.461538            0.282077            0.447324   
28    NDAR_INV5F2R3TL8       0.923077            0.315824            0.331288   
29    NDAR_INVVZR097LZ       0.615385            0.341652            0.403136   
...                ...            ...                 ...                 ...   
7772  NDAR_INVM889LV1R       0.923077            0.582699            0.446642   
7773  NDAR_INV4BRVTNUE       0.192308            0.283259            0.544364   
7774  NDAR_INV7MCB12JE       0.961538            0.208414            0.412791   
7775  NDAR_INVJ5B6E6MK       0.038462            0.244834            0.278830   
7776  NDAR_INVG5M25E6R       0.076923            0.385407            0.588593   
7777  NDAR_INVHYH97A58       0.923077            0.429242            0.374021   
7778  NDAR_INVRREA9RBW       0.230769            0.277660            0.411443   
7779  NDAR_INV4L1WDFJ7       0.461538            0.349091            0.407091   
7780  NDAR_INVP3LDF3U5       0.115385            0.253218            0.313146   
7781  NDAR_INV528FZWXG       0.961538            0.146972            0.226765   
7782  NDAR_INV5G3DP835       0.038462            0.248891            0.316784   
7783  NDAR_INVR9MG0HJ4       0.076923            0.381043            0.440471   
7784  NDAR_INVM70F0J7W       0.500000            0.287694            0.389008   
7785  NDAR_INV0YLLNXRL       0.076923            0.273874            0.352419   
7786  NDAR_INV02JXJUZG       0.461538            0.373613            0.497756   
7787  NDAR_INVYJ4PHV0Y       0.538462            0.423314            0.460990   
7788  NDAR_INV7GYDJXB0       0.884615            0.383425            0.673486   
7789  NDAR_INV0JRNB4U4       0.884615      

In [10]:
# Create correlation matrix
corr_matrix = all_data.corr().abs()

# Select upper triangle of correlation matrix
upper = corr_matrix.where(np.triu(np.ones(corr_matrix.shape), k=1).astype(np.bool))

In [11]:

# Find index of feature columns with correlation greater than 0.95
to_drop_dueTOhighCor = [column for column in upper.columns if any(upper[column] > 0.95)]
to_drop_dueTOhighCor

['SRI24THALAMUSRGM', 'SRI24CBLMHEMIWHTRWM', 'SRI24WM400WM400RWM']

In [12]:
all_data = all_data[all_data.columns.difference(to_drop_dueTOhighCor)].copy()

# checked for 0 variance columns, there are none.

___

### Splitting in train (80%), validation_internal (5%), validation_forEnsemble (15%) with strata on sex, IQ and age

In [13]:
# extracting columns on which I need to create strata 


trainyValy_moredt['age_bin']=pd.qcut(trainyValy_moredt['INTERVIEW_AGE'], 5, labels=list(range(0,5)))
trainyValy_moredt['Score_bin']=pd.qcut(trainyValy_moredt['residual_fluid_intelligence_score'], 10,\
                                       labels=list(range(0,10)))
trainyValy_moredt.STUDY_COHORT_NAME.value_counts()

ABCD NP Challenge 2019 Training Set      3736
ABCD NP Challenge 2019 Validation Set     415
Name: STUDY_COHORT_NAME, dtype: int64

In [14]:
val = trainyValy_moredt[trainyValy_moredt['STUDY_COHORT_NAME']=='ABCD NP Challenge 2019 Validation Set']
trainy_moredt = trainyValy_moredt[trainyValy_moredt['STUDY_COHORT_NAME']=='ABCD NP Challenge 2019 Training Set']
print(val.shape)
print(trainy_moredt.shape)


(415, 8)
(3736, 8)


In [15]:
# splitting training into 80-20 and the 20 further into 75-25 (75 for ensemble and 25 our internal validation)

train, vals = train_test_split(trainy_moredt, \
                                       test_size=0.20, \
                                       random_state=243,\
                                       stratify=trainy_moredt[['Score_bin', 'sex_bin','age_bin']])
val_ForEnsemble, val_internal = train_test_split(vals, \
                                       test_size=0.25, \
                                       random_state=243,\
                                       stratify=vals[['Score_bin', 'sex_bin','age_bin']])

In [16]:
print("train shape >",train.shape)
print("val_ForEnsemble shape >",val_ForEnsemble.shape)
print("val_internal shape >",val_internal.shape)

print(train.Score_bin.value_counts())
print(val_ForEnsemble.Score_bin.value_counts())
print(val_internal.Score_bin.value_counts())


train shape > (2988, 8)
val_ForEnsemble shape > (561, 8)
val_internal shape > (187, 8)
8    310
9    307
0    304
2    301
6    300
3    300
7    299
1    292
5    288
4    287
Name: Score_bin, dtype: int64
8    59
6    57
2    57
1    57
9    56
7    56
3    56
0    56
4    54
5    53
Name: Score_bin, dtype: int64
9    21
5    21
3    19
0    19
8    18
7    18
6    18
4    18
1    18
2    17
Name: Score_bin, dtype: int64


In [17]:
# adding category column
train['sample'] = 'train'
val_ForEnsemble['sample'] = 'val_forEnsemble'
val_internal['sample'] = 'val_internal'
val['sample'] = 'val'

/home/shikhar/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  
/home/shikhar/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  This is separate from the ipykernel package so we can avoid doing imports until
/home/shikhar/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value in

In [18]:
allY_withsegments = train.append(val_ForEnsemble).append(val_internal).append(val).copy()
allY_withsegments['sample'].value_counts()

train              2988
val_forEnsemble     561
val                 415
val_internal        187
Name: sample, dtype: int64

In [19]:
allY_withsegments.to_csv(data_dir+'allY_withsegments.csv', sep=',',index=False)


In [20]:
all_data = pd.merge(all_data, allY_withsegments[['residual_fluid_intelligence_score','SUBJECTKEY','sample']], on='SUBJECTKEY',how='left')



In [21]:
all_data['sample'].isna().value_counts()

False    4151
True     3651
Name: sample, dtype: int64

In [22]:
all_data.to_csv(data_dir+'all_volumetrix_processed.csv', sep=',',index=False)

INTERVIEW_AGE  SRI24AMYGDALALGM  SRI24AMYGDALARGM  SRI24ANGULARLGM  \
0          0.730769          0.414616          0.498913         0.408479   
1          0.653846          0.503110          0.495210         0.403832   
2          0.500000          0.362114          0.445087         0.258911   
3          0.384615          0.354227          0.495866         0.564634   
4          0.500000          0.504922          0.536298         0.415827   
5          0.076923          0.548186          0.623860         0.366376   
6          0.307692          0.282067          0.379114         0.295948   
7          0.500000          0.574076          0.609198         0.375601   
8          0.269231          0.350792          0.536245         0.365410   
9          0.384615          0.409298          0.545981         0.261631   
10         0.807692          0.274744          0.468093         0.465260   
11         0.884615          0.371269          0.507542         0.347618   
12         0.538462          0.278778          0.511156         0.564611   
13         0.346154          0.413953          0.375305         0.299900   
14         0.653846          0.381466          0.425172         0.472070   
15         0.730769          0.511661          0.573842         0.216586   
16         0.807692          0.471962          0.529441         0.502023   
17         0.884615          0.516718          0.591852         0.353390   
18         0.076923          0.573709          0.438221         0.723013   
19         0.923077          0.295884          0.489213         0.344857   
20         0.346154          0.369801          0.439949         0.268466   
21         0.692308          0.478156          0.572894         0.434483   
22         0.038462          0.393760          0.448267         0.522589   
23         0.230769          0.228389          0.413407         0.481274   
24         0.307692          0.482964          0.757700         0.434417   
25         0.230769          0.373839          0.482445         0.364401   
26         0.346154          0.540891          0.621149         0.664416   
27         0.461538          0.223142          0.259439         0.459847   
28         0.923077          0.597763          0.730264         0.126649   
29         0.615385          0.553125          0.505265         0.448497   
...             ...               ...               ...              ...   
7772       0.923077          0.480880          0.613627         0.228871   
7773       0.192308          0.505739          0.495990         0.273953   
7774       0.961538          0.352296          0.359988         0.256093   
7775       0.038462          0.558123          0.577288         0.523118   
7776       0.076923          0.387744          0.584499         0.455310   
7777       0.923077          0.351657          0.371930         0.472902   
7778       0.230769          0.885984          0.388310         0.674199   
7779       0.461538          0.505704          0.584172         0.344455   
7780       0.115385          0.239850          0.422718         0.265177   
7781       0.961538          0.683402          0.783878         0.223378   
7782       0.038462          0.332352          0.423462         0.247804   
7783       0.076923          0.259664          0.512822         0.379950   
7784       0.500000          0.485191          0.447355         0.599916   
7785       0.076923          0.330137          0.452289         0.498696   
7786       0.461538          0.336912          0.523249         0.659114   
7787       0.538462          0.577060          0.575889         0.359444   
7788       0.884615          0.388075          0.473116         0.562667   
7789       0.884615          0.425145          0.570511         0.442504   
7790       0.500000          0.280046          0.507391         0.561171   
7791       0.488542          0.440025          0.518134         0.417611   
7792       0.488542          0.440025          0.51813